# 1장 1강: 가설검정 프레임워크의 이해 — 실습문제

## 실습 목표

- 분석 상황에 맞게 귀무가설(H₀)과 대립가설(H₁)을 설정할 수 있다.
- `scipy.stats.ttest_1samp()`로 단일표본 t검정을 수행할 수 있다.
- 유의수준과 p-value를 비교하여 귀무가설 기각 여부를 판단할 수 있다.
- 제1종 오류와 제2종 오류를 분석 상황에 맞게 설명할 수 있다.
- 연구 목적에 따라 양측검정과 단측검정을 구분할 수 있다.

## 실습 환경 / 데이터

- Python
- pandas
- scipy.stats
- `ames_housing.csv`

이번 실습에서는 미국 아이오와주 Ames 지역의 주택 자료를 사용합니다.

주요 컬럼은 다음과 같습니다.

| 컬럼 | 의미 |
|---|---|
| `SalePrice` | 주택 판매가격 |
| `GrLivArea` | 지상층 생활면적 |
| `OverallQual` | 주택의 전반적인 품질 점수(1~10) |
| `YearBuilt` | 건축연도 |

> 모든 검정의 유의수준은 별도 지시가 없으면 `α = 0.05`로 설정합니다.

## 실습 준비

1. 필요한 라이브러리를 불러오세요.
2. Ames Housing 데이터를 `df`에 불러오세요.
3. 데이터의 행과 열 개수, 컬럼명, 상위 5개 행을 확인하세요.

In [ ]:
# 실습 준비 코드를 작성하세요.
from pathlib import Path
import pandas as pd
import numpy as np
from scipy import stats
df = pd.read_csv(r'C:\Machine_Learning_worksheet\Data\ames_housing.csv', encoding ='cp949')


---

## 필수 1. 평균 판매가격에 대한 양측검정

### 문제 1-1. 평균 판매가격이 180,000달러와 다른가?

#### 문제 설명

Ames 지역의 주택 평균 판매가격에 대한 기존 기준이 **180,000달러**라고 가정합니다.  
제공된 표본의 평균 판매가격이 이 기준과 다른지 단일표본 t검정으로 확인하세요.

#### 요구사항

1. `SalePrice`에서 결측치를 제거하여 `sale_price`에 저장하세요.
2. 표본 수와 표본평균을 출력하세요.
3. 다음 가설을 코드 주석이나 마크다운으로 작성하세요.
   - H₀: Ames 지역의 평균 판매가격은 180,000달러이다.
   - H₁: Ames 지역의 평균 판매가격은 180,000달러가 아니다.
4. 유의수준 `alpha`를 0.05로 설정하세요.
5. `stats.ttest_1samp()`와 `alternative="two-sided"`를 사용해 양측검정을 수행하세요.
6. t통계량과 p-value를 소수점 넷째 자리까지 출력하세요.
7. p-value와 유의수준을 비교해 귀무가설 기각 여부를 출력하세요.

#### 해석 질문

**Q1.** 이 문제에서 양측검정을 사용하는 이유는 무엇인가요?  
**Q2.** 검정 결과에 따라 귀무가설을 기각할 수 있나요?  
**Q3.** 귀무가설을 기각하지 못했다면 평균 판매가격이 정확히 180,000달러라고 증명된 것인가요?

#### 제출 결과

- 가설 설정
- 표본 수와 표본평균
- 단일표본 t검정 코드
- t통계량과 p-value
- 귀무가설 판단
- Q1~Q3 답변

In [8]:
# 필수 1 코드를 작성하세요.

sale_price = df['SalePrice'].dropna()
mu0 = 180000
alpha = 0.05

result = stats.ttest_1samp(
    sale_price,
    popmean=mu0,
    alternative ="two-sided"
)
print(f'표본 수 : {len(sale_price)}')
print(f'sample mean : {sale_price.mean(): ,.2f} dollars')
print(f'stat: {result.statistic: .4f}')
print(f'p-value: {result.pvalue: .4f}')

if result.pvalue <= alpha:
    print('judgement: 귀무가설을 기각합니다.')
else:
    print('judgement: 귀무가설을 기각하지 못합니다')

KeyError: 'SalePrice'

### 필수 1 답변 작성란

- **Q1.**이 문제에서 양측 검정을 사용하는 이유는 무엇인가요
평균이 기준보다 높은지 낮은지 방향성을 물어보는 것이 아니라 180,000달러와 다른 지를 묻는 문제이기 때문
- **Q2.** 검정 결과에 따라 귀무가설을 기각할 수 있나요?
기각할 수 없음 P-value 가 약 0.6578로 유의 수준 0.05보다 크다

- **Q3.** 귀무가설을 기각하지 못했다면 평균 판매 가격이 정확히 180,000 달러라고 증명된것인가?
아니요 귀무가설을 기각할 증거가 부족하다는 의미, 귀무가설이 참이라고 증명할 근거를 찾은 것이 아님


---

## 필수 2. 양측검정과 단측검정 비교

### 문제 2-1. 평균 품질 점수는 6점과 다른가, 그리고 6점보다 높은가?

#### 문제 설명

주택의 전반적인 품질을 나타내는 `OverallQual`의 기준 점수를 **6점**으로 설정했습니다.  
같은 표본으로 다음 두 연구 질문을 각각 검정하고 결과를 비교하세요.

- 질문 A: 평균 품질 점수가 6점과 **다른가?**
- 질문 B: 평균 품질 점수가 6점보다 **높은가?**

#### 요구사항

1. `OverallQual`에서 결측치를 제거하여 `overall_qual`에 저장하세요.
2. 표본 수와 표본평균을 출력하세요.
3. 질문 A의 가설을 작성하고 `alternative="two-sided"`로 양측검정을 수행하세요.
4. 질문 B의 가설을 작성하고 `alternative="greater"`로 단측검정을 수행하세요.
5. 두 검정의 t통계량과 p-value를 출력하세요.
6. 유의수준 0.05를 기준으로 각 검정의 귀무가설 기각 여부를 출력하세요.
7. 같은 데이터인데 두 p-value가 다르게 계산되는 이유를 설명하세요.

#### 해석 질문

**Q1.** 질문 A와 질문 B의 대립가설은 어떻게 다른가요?  
**Q2.** 양측검정과 단측검정의 결론은 각각 무엇인가요?  
**Q3.** 분석 결과를 먼저 확인한 뒤 p-value를 작게 만들기 위해 단측검정으로 바꿔도 되나요?  
**Q4.** 실제 평균 품질 점수가 6점인데도 검정 결과가 유의하다고 판단했다면 어떤 오류인가요?

#### 제출 결과

- 두 검정의 가설
- 표본 수와 표본평균
- 양측검정 및 단측검정 코드
- 두 검정의 t통계량과 p-value
- 검정별 귀무가설 판단
- Q1~Q4 답변

In [9]:
# 필수 2 코드를 작성하세요.



MU0 = 6       # 귀무가설 기준값
ALPHA = 0.05  # 유의수준
 
 
# ---------------------------------------------------------------
# 1. OverallQual 결측치 제거
# ---------------------------------------------------------------
overall_qual = df["OverallQual"].dropna()
 
 
# ---------------------------------------------------------------
# 2. 표본 수와 표본평균
# ---------------------------------------------------------------
n = len(overall_qual)
sample_mean = overall_qual.mean()
 
print("=" * 55)
print("[기초 통계]")
print(f"표본 수(n)  : {n}")
print(f"표본평균     : {sample_mean:.4f}")
print(f"표본표준편차 : {overall_qual.std(ddof=1):.4f}")
 
 
# ---------------------------------------------------------------
# 3. 질문 A - 양측검정
#    H0: mu = 6   (평균 품질 점수는 6점이다)
#    H1: mu != 6  (평균 품질 점수는 6점이 아니다 - 방향 지정 없음)
# ---------------------------------------------------------------
t_two, p_two = stats.ttest_1samp(overall_qual, popmean=MU0, alternative="two-sided")
 
 
# ---------------------------------------------------------------
# 4. 질문 B - 단측검정
#    H0: mu <= 6  (평균 품질 점수는 6점 이하다)
#    H1: mu >  6  (평균 품질 점수는 6점보다 높다 - 방향 지정)
# ---------------------------------------------------------------
t_one, p_one = stats.ttest_1samp(overall_qual, popmean=MU0, alternative="greater")
 
 
# ---------------------------------------------------------------
# 5. t통계량과 p-value 출력
# ---------------------------------------------------------------
print("=" * 55)
print("[검정 결과]")
print(f"질문 A (양측, mu != {MU0})  t = {t_two:.4f},  p = {p_two:.6f}")
print(f"질문 B (단측, mu >  {MU0})  t = {t_one:.4f},  p = {p_one:.6f}")
 
 
# ---------------------------------------------------------------
# 6. 유의수준 0.05 기준 귀무가설 기각 여부
# ---------------------------------------------------------------
print("=" * 55)
print(f"[판정] 유의수준 alpha = {ALPHA}")
 
for label, p in [("질문 A (양측검정)", p_two), ("질문 B (단측검정)", p_one)]:
    if p < ALPHA:
        print(f"{label}: p = {p:.6f} < {ALPHA} -> 귀무가설 기각")
    else:
        print(f"{label}: p = {p:.6f} >= {ALPHA} -> 귀무가설 기각 실패")
 
 
# ---------------------------------------------------------------
# 7. 같은 데이터인데 두 p-value가 다른 이유
# ---------------------------------------------------------------
print("=" * 55)
print("[7. 같은 데이터인데 p-value가 다른 이유]")
 
if t_two > 0:
    relation = (f"t > 0 이므로 두 꼬리가 대칭이라 정확히 단측 p = 양측 p / 2 가 된다.\n"
                f"실제로 {p_two:.6f} / 2 = {p_two / 2:.6f} 이고, 단측 p = {p_one:.6f} 로 일치한다.")
else:
    relation = (f"t < 0 이라 방향이 대립가설과 반대다. 이 경우 단측 p = 1 - 양측 p / 2 가 되어\n"
                f"오히려 양측보다 커진다 (단측 p = {p_one:.6f}).")
 
print(f"""
t통계량은 두 검정에서 완전히 동일하다 (t = {t_two:.4f}).
데이터가 같으니 당연하다. 달라지는 것은 t분포에서 '어느 영역의 넓이를
p-value로 세느냐'뿐이다.
 
- 양측검정: 대립가설이 방향을 지정하지 않으므로, 관측된 t만큼 극단적인
  경우를 오른쪽 꼬리와 왼쪽 꼬리 양쪽에서 센다.
    p = P(T >= |t|) + P(T <= -|t|)
 
- 단측검정: 대립가설이 '크다'는 방향을 미리 지정했으므로, 오른쪽 꼬리만 센다.
    p = P(T >= t)
 
{relation}
 
즉 p-value 차이는 데이터가 달라서가 아니라, 대립가설을 어떻게 세웠는지에
따라 '증거로 인정할 방향'이 달라졌기 때문이다.
""")
print("=" * 55)
 

[기초 통계]
표본 수(n)  : 1460
표본평균     : 6.0993
표본표준편차 : 1.3830
[검정 결과]
질문 A (양측, mu != 6)  t = 2.7439,  p = 0.006146
질문 B (단측, mu >  6)  t = 2.7439,  p = 0.003073
[판정] 유의수준 alpha = 0.05
질문 A (양측검정): p = 0.006146 < 0.05 -> 귀무가설 기각
질문 B (단측검정): p = 0.003073 < 0.05 -> 귀무가설 기각
[7. 같은 데이터인데 p-value가 다른 이유]

t통계량은 두 검정에서 완전히 동일하다 (t = 2.7439).
데이터가 같으니 당연하다. 달라지는 것은 t분포에서 '어느 영역의 넓이를
p-value로 세느냐'뿐이다.

- 양측검정: 대립가설이 방향을 지정하지 않으므로, 관측된 t만큼 극단적인
  경우를 오른쪽 꼬리와 왼쪽 꼬리 양쪽에서 센다.
    p = P(T >= |t|) + P(T <= -|t|)

- 단측검정: 대립가설이 '크다'는 방향을 미리 지정했으므로, 오른쪽 꼬리만 센다.
    p = P(T >= t)

t > 0 이므로 두 꼬리가 대칭이라 정확히 단측 p = 양측 p / 2 가 된다.
실제로 0.006146 / 2 = 0.003073 이고, 단측 p = 0.003073 로 일치한다.

즉 p-value 차이는 데이터가 달라서가 아니라, 대립가설을 어떻게 세웠는지에
따라 '증거로 인정할 방향'이 달라졌기 때문이다.



### 필수 2 답변 작성란

- **Q1.**
차이는 방향 지정 여부야. 질문 A의 대립가설은 H₁: μ ≠ 6 — 6점보다 높든 낮든 상관없이 "다르기만 하면" 지지 증거로 친다. 질문 B는 H₁: μ > 6 — 높은 쪽만 증거로 인정하고, 평균이 6보다 낮게 나오는 경우는 아무리 극단적이어도 기각 근거가 안 돼.

- **Q2.**
Ames train.csv 기준이면 n = 1460, 평균 ≈ 6.0993, t ≈ 2.74가 나와서 양측 p ≈ 0.0062, 단측 p ≈ 0.0031 — 둘 다 0.05 미만이라 양쪽 다 귀무가설 기각이야. 다만 결론 문장은 달라져야 해. 양측은 "6점과 다르다"까지만 말할 수 있고, 단측은 "6점보다 높다"고 방향까지 말할 수 있어.

- **Q3.**
안 돼. 이게 전형적인 p-hacking이야. 단측검정이 p를 반으로 줄여주는 건 "어느 방향으로 나올지 미리 걸었다"는 대가를 치렀기 때문인데, 결과를 본 뒤에 방향을 정하면 그 대가를 안 내고 이득만 챙기는 거지. 실제 1종 오류율이 0.05가 아니라 0.10이 돼버려. 검정 방향은 데이터를 보기 전에 도메인 지식이나 연구 질문에서 정해야 하고, 애매하면 양측이 기본값이야.
- **Q4.**
1종 오류(Type I error, 위양성). 귀무가설(μ = 6)이 실제로 참인데 기각한 경우니까. 유의수준 0.05는 "귀무가설이 참일 때 이런 실수를 할 확률을 5%까지 허용하겠다"는 선언이라, 이 오류는 검정을 잘못해서 생기는 게 아니라 설계상 일정 비율로 반드시 발생해.

---

## 과제 1. 평가 문항 기반 독립 과제

### 문제 3-1. 평균 지상층 생활면적이 1,550보다 작은가?

#### 문제 설명

한 부동산 분석가는 Ames 지역 주택의 평균 지상층 생활면적이 **1,550보다 작다**고 주장합니다.  
`GrLivArea`를 사용하여 이 주장을 단일표본 t검정으로 확인하세요.

> 이 문제는 필수 문제에서 연습한 가설 설정과 단측검정 절차를 새로운 변수에 적용하는 문제입니다.

#### 요구사항

1. `GrLivArea`에서 결측치를 제거하여 분석 표본을 준비하세요.
2. 표본 수와 표본평균을 확인하세요.
3. 연구 주장에 맞는 귀무가설과 대립가설을 작성하세요.
4. 양측검정과 단측검정 중 적절한 검정 방향을 선택하고 그 이유를 설명하세요.
5. 단일표본 t검정을 수행하세요.
6. t통계량과 p-value를 소수점 넷째 자리까지 출력하세요.
7. 유의수준 0.05를 기준으로 귀무가설 기각 여부를 판단하세요.
8. 분석가의 주장을 뒷받침할 통계적 근거가 있는지 해석하세요.

#### 해석 질문

**Q1.** 이 문제에서 적절한 `alternative` 값은 무엇인가요?  
**Q2.** p-value는 귀무가설이 참일 확률을 의미하나요?  
**Q3.** 실제 평균이 1,550보다 작은데 귀무가설을 기각하지 못했다면 어떤 오류인가요?

#### 제출 결과

- 가설 설정과 검정 방향 선택 이유
- 표본 수와 표본평균
- 검정 코드
- t통계량과 p-value
- 귀무가설 판단과 분석 결론
- Q1~Q3 답변

In [2]:
# 과제 1 코드를 작성하세요.

from scipy import stats

# 1) 결측치 제거
x = df['GrLivArea'].dropna()

# 2) 표본 수와 표본평균
n, xbar = len(x), x.mean()
print(f"표본 수 n = {n}")
print(f"표본평균 = {xbar:.4f}")

# 5) 단일표본 t검정 (좌측 단측)
t_stat, p_value = stats.ttest_1samp(x, popmean=1550, alternative='less')

# 6) 소수점 넷째 자리까지 출력
print(f"t-통계량 = {t_stat:.4f}")
print(f"p-value  = {p_value:.4f}   (지수표기: {p_value:.4e})")

# 7) 유의수준 0.05 기준 판단
alpha = 0.05
if p_value < alpha:
    print(f"p-value < {alpha} → 귀무가설 기각")
else:
    print(f"p-value >= {alpha} → 귀무가설 기각 실패")

표본 수 n = 1460
표본평균 = 1515.4637
t-통계량 = -2.5113
p-value  = 0.0061   (지수표기: 6.0683e-03)
p-value < 0.05 → 귀무가설 기각


### 과제 1 답변 작성란

**Q1.** 이 문제에서 적절한 `alternative` 값은 무엇인가요?  
적절한 alternative 값은 Less 이다 왜냐하면 
생활면적이 **1,550보다 작다**고 주장을 하고 있기 때문에. 

**Q2.** p-value는 귀무가설이 참일 확률을 의미하나요? 

아니요. 
p-value 는 데이터를 보고 이것이 참인지 아닌지를 판단하는 도구가 아님


**Q3.** 실제 평균이 1,550보다 작은데 귀무가설을 기각하지 못했다면 어떤 오류인가요?
Type II error로 봐야한다. 

왜냐하면 
μ < 1550 (H₀ 거짓) 이부분에서 못 쳐내서 잘못(실제로 마지막에 단계에 부분 확인)
→ 그런데 표본에서 계산한 p-value가 0.05 이상으로 나옴
→ 규칙대로 "기각 못 함" 판정


---

## 실습 마무리

아래 질문에 답하세요.

1. 이번 실습에서는 표본평균이 기준값과 다른지 판단하기 위해 어떤 검정을 사용했나요?
-> 단일 표본 T검정, 하나의 표본을 이용하여 모집단 평균이 특정 기준값과 다른지
-> 또는 한쪽 방향으로 차이가 있는지 확인
-> 1장 1강 실습문제에서는 주택 판매가격 표본을 이용하여 평균이 180,000과 다른지 검정햇다

2. 귀무가설의 기각 여부는 무엇을 기준으로 판단했나요?
-> P-value와 사전에 설정한 유의수준(0.05)를 비교
-> p-value <= 0.05이면 귀무가설을 기각하고, p-value > 0.05면 기각하지 못함

3. 양측검정과 단측검정은 연구 질문에서 어떤 차이가 있나요?
-> 양측검정은 기준과 다르지 확인하고
-> 단측검정은 기준보다 크거나 작은 특정 방향을 확인한다.

4. 귀무가설을 기각하지 못한다는 것은 무엇을 의미하나요?
귀무가설을 참이라고 증명한 것이 아니라 대립가설을 지지할 증거 충분하지 않다는 의미입니다.

5. 제1종 오류와 제2종 오류는 각각 어떤 상황인가요?
-> 제 1종 오류는 실제로 귀무가설이 참인데 이를 기각하는 오류
-> 제 2종 오류는 실제로 귀무가설이 거짓인데 이를 기각하지 못하는 오류
-> 
